robottelo/report_portal/portal.py

```diff
@@ -0,0 +1,216 @@
+import logging
+
+import requests
+from tenacity import retry
+from tenacity import stop_after_attempt
+from tenacity import wait_fixed
+
+from robottelo.config import settings
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ReportPortal:
+    """Represents Report Portal in ReportPortal
+
+    This holds the properties and function to interact with the Report Portal
+    """
+
+    def __init__(self):
+        """Configure the settings and initiate report portal properties"""
+        if not settings.configured:
+            settings.configure()
+        self.rp_url = settings.report_portal.rp_url
+        self.rp_project = settings.report_portal.rp_project
+        self.rp_api_key = settings.report_portal.rp_key
+
+    @property
+    def api_url(self):
+        """Super url of report portal to be used as a base in all child classes
+        :returns Base url for API request
+        """
+        _api_url = f'{self.rp_url}/api/v1/{self.rp_project}/'
+        return _api_url
+
+    @property
+    def headers(self):
+        """The headers for Report Requests.
+        :returns header for API request
+        """
+        _headers = {'Authorization': f'Bearer {self.rp_api_key}'}
+        return _headers
+
+
+class Launch(ReportPortal):
```

Comment:
I don't think subclassing the Launch and tests from each other is the best approach here.

This might get called over-engineering it, but I think a different approach would better represent how these entities are related.

In my view, `launches` should be a function on the `ReportPortal` instance, not a property of the `Launch` class itself.

A report portal has launches, those launches have tests - this suggests a composited structure to me, not inherited.

`ReportPortal` instance attributes are correct as they are, and the `Launch` instance attribute is correct, but their relationship is not accurately represented to me, in this structure.

I'm not suggesting we go as far as concerning ourselves with whether we are collecting against slightly stale data, and completely over-engineering the request for launch data, but that we re-structure these a bit.

- `ReportPortal.launches` should be returning a data structure of Launch instances (keyed on some launch metadata?)
- `ReportPortal.launch` should be returning a specific launch based on passed arg (version + snap number)
- `Launch.tests` should be returning `Test` class instances
- `Test.launch` should be an instance attribute passed to its constructor from the Launch instance that it belongs to
- `Launch.report_portal` should be an instance attribute passed to its constructor from the ReportPortal instance it belongs to

I think that the current `Tests` class is misleading, given the url instance attribute is for an item.

We can simply subclass the `Test` class from AttrDict, instantiating it with the processed JSON from the report portal response, so that plugins using this ReportPortal tool are able to get a list of test items, and attributize the metadata comparisons needed for test collection.

This creates a pattern where in the plugin caller, we're doing something that looks like:
```
rp = ReportPortal()
current_version = pretend_with_me_method_that_gets_current_test_version()
# current_version  is something like '6.7.0-20.0' for snap 20
# assume pytest plugin hook, we have list of collected tests
test_items = complete_collected_tests
already_executed_tests = rp.launch(current_version).tests  # dict keyed on test id
tests_to_rerun = []
tests_that_passed = []
for test in complete_collected_tests:
    if already_executed_tests.get(test.id):
        if already_executed_tests[test.id].status.overall == 'passed':
            tests_that_passed.append(test)
        else:
            tests_to_rerun.append(test)
```

Reply:
Yeah, while working with the collecter function, I observed some of the drawbacks of this design. Need to rework. I shall update the PR soon.

Reply:
The main intention behind subclassing them is to independently import those classes elsewhere. like:
```from robottelo.report_portal.portal import Tests``` 
where I need not to instantite Reportportal class. I can go independentaly.

Reply:
@mshriver , 

I am collecting tests to rerun like this (using current RP implmentation):
```
launch_tests_ = Tests(launch=launch_position)
rerunable_defect_types = ['no_defect', 'to_investigate', 'saucelabs_issue', 'system_issue']
for dtype in eligible_defect_types:
     rerunable_tests += [test['name'] for test in launch_tests_.tests(status='FAILED', defect_type=dtype)]
return rerunable_tests
```

I am only collecting tests (with their name) to rerun, Not sure why you are collecting passed tests above.

Reply:
I'm not seeing the advantage or intent when importing a `Launch` or `Test` class that is subclassed, and where you would not be instantiating a `ReportPortal` instance. Can you elaborate there?  I'd assume you will always need/want the ReportPortal instance for the other classes to be relevant.

We discussed the keep/discard test item collection intent in the call today, but to summarize:

By tracking both tests to keep in the collection and tests to discard, we can report to pytest collected cases which were discarded, which is useful in assessing test collection for a session.  We can also hook to report these discard statistics.

I also made the initial comment before reading your method for test listing, which includes direct filtering of results - so my example is unnecessarily iterating over the results to filter based on overall status.